### PR TITLE
Ho parsing and regressions

### PR DESCRIPTION
**Pull request by ajreynol**
This pull request:
- Adds parsing support for higher-order operators (lambda terms and function types). This is the last step for making CVC4 operational for higher-order constraints,
- Refactors smt2 parsing of define-fun-rec.  Previously, portions of the code for parsing define-fun-rec and define-funs-rec were copy and pasted in 3 places in Smt.g. Since parsing define-fun-rec require non-trivial changes in the presence of function return types, instead of making these changes in 3 places, I collected these into a new function in the smt2 parser.  This will also make #839 and #1316 easier to address.
- Adds regressions for higher-order. Currently, several regressions are not working (they rely on #1204 being merged).  These will be enabled once that pull request is merged.

### CODE REVIEW

**Comment by ajreynol** [comment]
This commit is 10/10 planned commits from the ho-smt branch: https://github.com/CVC4/CVC4/tree/ho-smt

File: src/parser/parser.cpp [hunk 1]
```diff
@@ -448,6 +448,64 @@ std::vector<DatatypeType> Parser::mkMutualDatatypeTypes(
...
 
+Type Parser::mkFlatFunctionType(std::vector<Type>& sorts,
+                                Type range,
+                                std::vector<Expr>& flattenVars)
+{
+  if (range.isFunction())
+  {
+    std::vector<Type> domainTypes =
+        (static_cast<FunctionType>(range)).getArgTypes();
+    for (unsigned i = 0; i < domainTypes.size(); i++)
```
Comment (HanielB) [inline src/parser/parser.cpp]: no ``size()`` in condition

File: src/parser/parser.cpp [hunk 2]
```diff
@@ -448,6 +448,64 @@ std::vector<DatatypeType> Parser::mkMutualDatatypeTypes(
...
+      Expr v = d_exprManager->mkBoundVar(ss.str(), domainTypes[i]);
+      flattenVars.push_back(v);
+    }
+    range = static_cast<FunctionType>(range).getRangeType();
+  }
+  if (sorts.empty())
+  {
+    return range;
+  }
+  else
```
Comment (HanielB) [inline src/parser/parser.cpp]: No-else-after-return

File: src/parser/parser.cpp [hunk 3]
```diff
@@ -448,6 +448,64 @@ std::vector<DatatypeType> Parser::mkMutualDatatypeTypes(
...
+}
+
+Type Parser::mkFlatFunctionType(std::vector<Type>& sorts, Type range)
+{
+  if (sorts.empty())
+  {
+    // no difference
+    return range;
+  }
+  else
```
Comment (HanielB) [inline src/parser/parser.cpp]: ditto

File: src/parser/parser.cpp [hunk 4]
```diff
@@ -448,6 +448,64 @@ std::vector<DatatypeType> Parser::mkMutualDatatypeTypes(
...
+      sorts.insert(sorts.end(), domainTypes.begin(), domainTypes.end());
+      range = static_cast<FunctionType>(range).getRangeType();
+    }
+    return d_exprManager->mkFunctionType(sorts, range);
+  }
+}
+
+Expr Parser::mkHoApply(Expr expr, std::vector<Expr>& args, unsigned startIndex)
+{
+  for (unsigned i = startIndex; i < args.size(); i++)
```
Comment (HanielB) [inline src/parser/parser.cpp]: for-each

File: src/parser/parser.cpp [hunk 5]
```diff
@@ -448,6 +448,64 @@ std::vector<DatatypeType> Parser::mkMutualDatatypeTypes(
...
+  {
+    expr = d_exprManager->mkExpr(kind::HO_APPLY, expr, args[i]);
```
Comment (HanielB) [inline src/parser/parser.cpp]: You are using ``CVC4::kind`` namespace, remove the prefix.

File: src/parser/parser.h [hunk 6]
```diff
@@ -573,6 +573,68 @@ class CVC4_PUBLIC Parser {
   std::vector<DatatypeType>
   mkMutualDatatypeTypes(std::vector<Datatype>& datatypes, bool doOverload=false);
 
+  /** mkFlatFunctionType
+   * Returns the "flat" function type correspond to the function taking argument
```
Comment (HanielB) [inline src/parser/parser.h]: newline
correspond --> corresponding?

File: src/parser/parser.h [hunk 7]
```diff
@@ -573,6 +573,68 @@ class CVC4_PUBLIC Parser {
   std::vector<DatatypeType>
   mkMutualDatatypeTypes(std::vector<Datatype>& datatypes, bool doOverload=false);
 
+  /** mkFlatFunctionType
```
Comment (HanielB) [inline src/parser/parser.h]: better brief

File: src/parser/parser.h [hunk 8]
```diff
@@ -573,6 +573,68 @@ class CVC4_PUBLIC Parser {
...
+   *    (define-fun Q ((x Int) (z Int)) Int (@ (lambda y (P x)) z))
+   * where @ is (higher-order) application. In this example, z is added to
+   * flattenVars.
+   */
+  Type mkFlatFunctionType(std::vector<Type>& sorts,
+                          Type range,
+                          std::vector<Expr>& flattenVars);
+
+  /** mkFlatFunctionType
+   * Same as above, but does not take argument flattenVars.
```
Comment (HanielB) [inline src/parser/parser.h]: newline

File: src/parser/parser.h [hunk 9]
```diff
@@ -573,6 +573,68 @@ class CVC4_PUBLIC Parser {
...
+   * This is used when the arguments of the function are not important (for
+   * instance, if we are only using this type in a declare-fun).
+   */
+  Type mkFlatFunctionType(std::vector<Type>& sorts, Type range);
+
+  /** mkHoApply
+  * This returns the left-associative curried application of (function) expr to
```
Comment (HanielB) [inline src/parser/parser.h]: ditto

File: src/parser/parser.h [hunk 10]
```diff
@@ -573,6 +573,68 @@ class CVC4_PUBLIC Parser {
...
+  * the arguments in args, starting at index startIndex.
+  *
+  * For example, mkHoApply( f, { a, b }, 0 ) returns
+  *  (HO_APPLY (HO_APPLY f a) b)
+  *
+  * If args is non-empty, the type of expr should be (-> T0 ... Tn T),
+  * where args[i-startIndex].getType() = Ti for each i where
+  * startIndex <= i < args.size(), otherwise the returned expression
+  * will not be well-typed.
+  */
```
Comment (HanielB) [inline src/parser/parser.h]: clarify

File: src/parser/smt2/Smt2.g [hunk 11]
```diff
@@ -383,7 +383,14 @@ command [std::unique_ptr<CVC4::Command>* cmd]
       }
     }
     term[expr, expr2]
-    { PARSER_STATE->popScope();
+    {
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: there is an empty space after the curly brace. Please remove it.

File: src/parser/smt2/Smt2.g [hunk 12]
```diff
@@ -1191,37 +1205,16 @@ smt25Command[std::unique_ptr<CVC4::Command>* cmd]
...
-        for(std::vector<std::pair<std::string, CVC4::Type> >::const_iterator i =
-            sortedVarNames.begin(), iend = sortedVarNames.end(); i != iend;
-            ++i) {
-          sorts.push_back((*i).second);
-        }
-        t = EXPR_MANAGER->mkFunctionType(sorts, t);
-      }
-      // allow overloading
-      Expr func = PARSER_STATE->mkVar(fname, t, ExprManager::VAR_FLAG_NONE, true);
+    {
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 13]
```diff
@@ -1246,23 +1239,19 @@ smt25Command[std::unique_ptr<CVC4::Command>* cmd]
...
-              sorts.push_back((*i).second);
-            }
-            t = EXPR_MANAGER->mkFunctionType(sorts, t);
-          }
-        }
-        sortedVarNames.clear();
-        // allow overloading
-        Expr func = PARSER_STATE->mkVar(fname, t, ExprManager::VAR_FLAG_NONE, true);
+      { 
+        flattenVars.clear();
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 14]
```diff
@@ -1246,23 +1239,19 @@ smt25Command[std::unique_ptr<CVC4::Command>* cmd]
...
+        func = PARSER_STATE->mkDefineFunRec( fname, sortedVarNames, t, flattenVars );
         seq->addCommand(new DeclareFunctionCommand(fname, func, t));
         funcs.push_back( func );
+        
+        // add to lists (need to remember for when parsing the bodies)
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 15]
```diff
@@ -1246,23 +1239,19 @@ smt25Command[std::unique_ptr<CVC4::Command>* cmd]
...
+        sortedVarNamesList.push_back( sortedVarNames );
+        flattenVarsList.push_back( flattenVars );
+        
+        // set up parsing the next variable list block
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 16]
```diff
@@ -1274,27 +1263,19 @@ smt25Command[std::unique_ptr<CVC4::Command>* cmd]
...
-              i = sortedVarNamesList[0].begin(),
-              iend = sortedVarNamesList[0].end(); i != iend; ++i) {
-          Expr v = PARSER_STATE->mkBoundVar((*i).first, (*i).second);
-          bvs.push_back( v );
-          f_app.push_back( v );
-        }
-        func_app = MK_EXPR( kind::APPLY_UF, f_app );
-      }
+      PARSER_STATE->pushDefineFunRecScope( sortedVarNamesList[0], funcs[0], 
+                                           flattenVarsList[0], func_app, bvs, true);
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 17]
```diff
@@ -1311,23 +1292,9 @@ smt25Command[std::unique_ptr<CVC4::Command>* cmd]
...
-                  i = sortedVarNamesList[j].begin(),
-                  iend = sortedVarNamesList[j].end(); i != iend; ++i) {
-            Expr v = PARSER_STATE->mkBoundVar((*i).first, (*i).second);
-            bvs.push_back( v );
-            f_app.push_back( v );
-          }
-          func_app = MK_EXPR( kind::APPLY_UF, f_app );
-        }
+        PARSER_STATE->pushDefineFunRecScope( sortedVarNamesList[j], funcs[j], 
+                                             flattenVarsList[j], func_app, bvs, true);
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 18]
```diff
@@ -2008,7 +1978,18 @@ termNonVariable[CVC4::Expr& expr, CVC4::Expr& expr2]
       if(isBuiltinOperator) {
         PARSER_STATE->checkOperator(kind, args.size());
       }
-      expr = MK_EXPR(kind, args); 
+      // may be partially applied function, in this case we should use HO_APPLY     
+      if( args.size()>=2 && args[0].getType().isFunction() &&
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 19]
```diff
@@ -2008,7 +1978,18 @@ termNonVariable[CVC4::Expr& expr, CVC4::Expr& expr2]
...
+          (args.size()-1)<((FunctionType)args[0].getType()).getArity() ){
+        Debug("parser") << "Partial application of " << args[0];
+        Debug("parser") << " : #argTypes = " << ((FunctionType)args[0].getType()).getArity();  
+        Debug("parser") << ", #args = " << args.size()-1 << std::endl;
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 20]
```diff
@@ -2008,7 +1978,18 @@ termNonVariable[CVC4::Expr& expr, CVC4::Expr& expr2]
...
+      if( args.size()>=2 && args[0].getType().isFunction() &&
+          (args.size()-1)<((FunctionType)args[0].getType()).getArity() ){
+        Debug("parser") << "Partial application of " << args[0];
+        Debug("parser") << " : #argTypes = " << ((FunctionType)args[0].getType()).getArity();  
+        Debug("parser") << ", #args = " << args.size()-1 << std::endl;  
+        // must curry the application
+        expr = args[0];
+        expr = PARSER_STATE->mkHoApply( expr, args, 1 );
+      }else{
+        expr = MK_EXPR(kind, args);
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 21]
```diff
@@ -2270,6 +2251,27 @@ termNonVariable[CVC4::Expr& expr, CVC4::Expr& expr2]
         expr2 = f2;
       }
     }
+  | /* lambda */
+    LPAREN_TOK HO_LAMBDA_TOK
+    LPAREN_TOK sortedVarList[sortedVarNames] RPAREN_TOK
+    {
+      PARSER_STATE->pushScope(true);
+      for(std::vector<std::pair<std::string, CVC4::Type> >::const_iterator i =
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: for-each

File: src/parser/smt2/Smt2.g [hunk 22]
```diff
@@ -2877,6 +2879,16 @@ sortSymbol[CVC4::Type& t, CVC4::parser::DeclarationCheck check]
         }
       }
     ) RPAREN_TOK
+  | LPAREN_TOK HO_ARROW_TOK sortList[args] RPAREN_TOK
+    {
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: trailing space

File: src/parser/smt2/smt2.cpp [hunk 23]
```diff
@@ -348,6 +348,69 @@ Expr Smt2::getExpressionForNameAndType(const std::string& name, Type t) {
...
 }
 
+Expr Smt2::mkDefineFunRec(
+    const std::string& fname,
+    const std::vector<std::pair<std::string, Type> >& sortedVarNames,
+    Type t,
+    std::vector<Expr>& flattenVars)
+{
+  std::vector<Type> sorts;
+  for (std::vector<std::pair<std::string, CVC4::Type> >::const_iterator
```
Comment (HanielB) [inline src/parser/smt2/smt2.cpp]: for-each

File: src/parser/smt2/smt2.cpp [hunk 24]
```diff
@@ -348,6 +348,69 @@ Expr Smt2::getExpressionForNameAndType(const std::string& name, Type t) {
...
+    std::vector<Expr>& bvs,
+    bool bindingLevel)
+{
+  pushScope(bindingLevel);
+
+  std::vector<Expr> f_app;
+  f_app.push_back(func);
+  // bound variables are those that are explicitly named in the preamble
+  // of the define-fun(s)-rec command, we define them here
+  for (std::vector<std::pair<std::string, CVC4::Type> >::const_iterator
```
Comment (HanielB) [inline src/parser/smt2/smt2.cpp]: for-each

File: src/parser/smt2/smt2.h [hunk 25]
```diff
@@ -91,6 +91,52 @@ class Smt2 : public Parser {
    */
   virtual Expr getExpressionForNameAndType(const std::string& name, Type t);
 
+  /** Make function defined by a define-fun(s)-rec command.
+  *
+  * fname : the name of the function.
+  * sortedVarNames : the list of variable arguments for the function.
+  * t : the range type of the function we are defining.
+  *
+  * This function will create a bind a new function term
```
Comment (HanielB) [inline src/parser/smt2/smt2.h]: hard to read

File: src/parser/smt2/smt2.h [hunk 26]
```diff
@@ -91,6 +91,52 @@ class Smt2 : public Parser {
...
+  * range type.
+  */
+  Expr mkDefineFunRec(
+      const std::string& fname,
+      const std::vector<std::pair<std::string, Type> >& sortedVarNames,
+      Type t,
+      std::vector<Expr>& flattenVars);
+
+  /** Push scope for define-fun-rec
+  * This calls Parser::pushScope(bindingLevel) and sets up
```
Comment (HanielB) [inline src/parser/smt2/smt2.h]: newline

File: src/parser/parser.cpp (hunk 4 above)
Comment (ajreynol) [reply src/parser/parser.cpp]: need for here because of startIndex, I'll make it size cached though

Comment (ajreynol) [reply src/parser/parser.cpp]: need for here because of startIndex, I'll make it size cached though

**Pull request merged**